includes/class-wc-payments-apple-pay-registration.php

```diff
@@ -0,0 +1,343 @@
+		$well_known_dir = untrailingslashit( ABSPATH ) . '/' . self::DOMAIN_ASSOCIATION_FILE_DIR;
+		$fullpath       = $well_known_dir . '/' . self::DOMAIN_ASSOCIATION_FILE_NAME;
+
+		if ( ! file_exists( $well_known_dir ) ) {
+			if ( ! @mkdir( $well_known_dir, 0755 ) ) { // @codingStandardsIgnoreLine
+				return __( 'Unable to create domain association folder to domain root.', 'woocommerce-payments' );
+			}
+		}
+
+		if ( ! @copy( WCPAY_ABSPATH . '/' . self::DOMAIN_ASSOCIATION_FILE_NAME, $fullpath ) ) { // @codingStandardsIgnoreLine
+			return __( 'Unable to copy domain association file to domain root.', 'woocommerce-payments' );
+		}
+	}
+
+	/**
+	 * Updates the Apple Pay domain association file.
+	 * Reports failure only if file isn't already being served properly.
+	 */
+	public function update_domain_association_file() {
+		if ( $this->is_hosted_domain_association_file_up_to_date() ) {
+			return;
+		}
+
+		$error_message = $this->copy_and_overwrite_domain_association_file();
+
+		if ( isset( $error_message ) ) {
+			$url = get_site_url() . '/' . self::DOMAIN_ASSOCIATION_FILE_DIR . '/' . self::DOMAIN_ASSOCIATION_FILE_NAME;
+			Logger::log(
+				'Error: ' . $error_message . ' ' .
+				/* translators: expected domain association file URL */
+				sprintf( __( 'To enable Apple Pay, domain association file must be hosted at %s.', 'woocommerce-payments' ), $url )
+			);
+		} else {
+			Logger::log( __( 'Domain association file updated.', 'woocommerce-payments' ) );
+		}
+	}
+
+	/**
+	 * Adds a rewrite rule for serving the domain association file from the proper location.
+	 */
+	public function add_domain_association_rewrite_rule() {
+		$regex    = '^\\' . self::DOMAIN_ASSOCIATION_FILE_DIR . '\/' . self::DOMAIN_ASSOCIATION_FILE_NAME . '$';
+		$redirect = 'index.php?' . self::DOMAIN_ASSOCIATION_FILE_NAME . '=1';
+
+		add_rewrite_rule( $regex, $redirect, 'top' );
+	}
+
+	/**
+	 * Add to the list of publicly allowed query variables.
+	 *
+	 * @param  array $query_vars - provided public query vars.
+	 * @return array Updated public query vars.
+	 */
+	public function whitelist_domain_association_query_param( $query_vars ) {
+		$query_vars[] = self::DOMAIN_ASSOCIATION_FILE_NAME;
+		return $query_vars;
+	}
+
+	/**
+	 * Serve domain association file when proper query param is provided.
+	 *
+	 * @param object $wp WordPress environment object.
+	 */
+	public function parse_domain_association_request( $wp ) {
+		if (
+			! isset( $wp->query_vars[ self::DOMAIN_ASSOCIATION_FILE_NAME ] ) ||
+			'1' !== $wp->query_vars[ self::DOMAIN_ASSOCIATION_FILE_NAME ]
+		) {
+			return;
+		}
+
+		$path = WCPAY_ABSPATH . '/' . self::DOMAIN_ASSOCIATION_FILE_NAME;
+		header( 'Content-Type: text/plain;charset=utf-8' );
+		echo esc_html( @file_get_contents( $path ) ); // @codingStandardsIgnoreLine
+		exit;
+	}
+
+	/**
+	 * Processes the Apple Pay domain verification.
+	 *
+	 * @throws API_Exception Exception thrown on registration failure.
```

Comment:
Since we catch `API_Exception` in this function I don't think we ever `throw` it, so let's remove this annotation 🙂

Reply:
Fixed it! Thanks again for the thorough review.